server/middleware/src/spaceSecurity.ts

```diff
@@ -400,6 +405,54 @@ export class SpaceSecurityMiddleware extends BaseMiddleware implements Middlewar
     return findResult
   }
 
+  override async searchFulltext (
+    ctx: SessionContext,
+    query: FulltextQuery,
+    options: FulltextQueryOptions
+  ): Promise<FulltextSearchResult> {
+    const newQuery = justClone(query)
+    const account = await getUser(this.storage, ctx)
+    let spaces: string[] = []
+    if (!isSystem(account)) {
+      spaces = await this.getAllAllowedSpaces(account)
+      const spacesFilter = {
```

Comment:
I'm not happy with unknown Elastic like syntax here, we need a structured/typed query and it should be handled inside adapter itself.

Reply:
I have looked at https://github.com/elastic/elasticsearch-js/blob/main/src/api/types.ts and thought that having all the possible typing for query it a bit too much. Maybe I wrong and we should take it? Or we could say that we only want to support some subset of what is possible, but I think it's against initial idea to be able to use full power of Elastic with specific request. 

Or I solve it quick way by passing spaces to adapter and the query will be not so typed, but the merge of spaces inside a query would happen in adapter and not here. 

Which approach do you like more?

Reply:
For me search API looks like following:

`search(mySearchString, options): SearchResult[]` and thats all.

Options could contain a limit on documents we would like to search, groupping and limits per individual classes etc.